mqlc/mqlc.go

```diff
@@ -1684,18 +1684,111 @@ func (c *compiler) postCompile() {
+					c.addValueFieldChunks(ref)
+				}
 			default:
 				c.expandResourceFields(chunk, typ, ref)
 			}
 		}
 	}
 }
 
+// addValueFieldChunks takes the value fields of the assessment and adds them to the
+// block forthe default fields
+// This way, the actual data of the assessment automatically shows up in the output
+// of the assessment that failed the assessment
+func (c *compiler) addValueFieldChunks(ref uint64) {
+	var assessmentBlock *llx.Block
+	var whereChunk *llx.Chunk
+
+	// find chunk with where/whereNot function
+	for {
+		chunk := c.Result.CodeV2.Chunk(ref)
+		if chunk.Function == nil {
+			// this is a safe guard and shouldn't happen
+			log.Error().Msg("failed to find where function for assessment")
+			return
+		}
+		if chunk.Function.Args != nil {
+			whereChunk = chunk
+			break
+		}
+		ref = chunk.Function.Binding
+	}
+
+	// find the referenced blcok for the where function
+	for i := len(whereChunk.Function.Args) - 1; i >= 0; i-- {
+		arg := whereChunk.Function.Args[i]
+		if types.Type(arg.Type).Underlying() == types.FunctionLike {
+			raw := arg.RawData()
+			blockRef := raw.Value.(uint64)
+			assessmentBlock = c.Result.CodeV2.Block(blockRef)
+			break
+		}
+	}
+
+	defaultFieldsBlock := c.Result.CodeV2.Blocks[len(c.Result.CodeV2.Blocks)-1]
+	defaultFieldsRef := defaultFieldsBlock.Chunks[1].Function.Binding
+
+	chunksToAdd := []*llx.Chunk{}
+	// Check whether the chunk is already present in the default fields block
+	// This can happen, when the assessment checks one of the default fields
+	// We can skip the first Chunk, as it is only the resource binding
+	for i := 1; i < len(assessmentBlock.Chunks)-1; i++ {
+		chunk := assessmentBlock.Chunks[i]
+		// filter out nested function block and only add fields for outer most resource
+		if strings.HasPrefix(chunk.Id, "$where") {
+			break
+		}
+		chunkAlreadyPresent := false
+		for j := range defaultFieldsBlock.Chunks {
+			if chunk.Id == defaultFieldsBlock.Chunks[j].Id {
+				chunkAlreadyPresent = true
+				break
+			}
+		}
+		if chunkAlreadyPresent {
+			continue
+		}
+		newChunk := &llx.Chunk{
+			Call: chunk.Call,
+			Id:   chunk.Id,
+		}
+		if chunk.Function != nil {
+			newChunk.Function = &llx.Function{
+				Binding: chunk.Function.Binding,
+				Type:    chunk.Function.Type,
+				Args:    chunk.Function.Args,
+			}
+		}
+		chunksToAdd = append(chunksToAdd, newChunk)
+	}
+	if len(chunksToAdd) == 0 {
+		return
+	}
+	chunkRef := defaultFieldsRef
+	for i := range chunksToAdd {
+		newChunk := chunksToAdd[i]
+		if newChunk.Function != nil {
+			newChunk.Function.Binding = chunkRef
+		}
+
+		defaultFieldsBlock.AddChunk(c.Result.CodeV2, chunkRef, newChunk)
+		chunkRef = defaultFieldsBlock.TailRef(chunkRef)
+	}
+	// Only add last chunk as entrypoint, to not get data multiple times
+	// e.g. for dict["key"] we would otherwise get the entry and the whole dict
+	defaultFieldsBlock.Entrypoints = append(defaultFieldsBlock.Entrypoints, chunkRef)
```

Comment:
This doesn't seem right to me. If you dont have each thing being collected as a entrypoint, its going to get skipped. That led me to try this example:
```
dlv debug --headless --api-version=2 --listen=:2345 apps/cnquery/cnquery.go -- run local --ast -c 'users.all(sshkeys.length > 2 && enabled == true)'
```

```
-> block 1
   entrypoints: [<1,5>]
   1: users
   2: list bind: <1,1> type:[]user
   3: $whereNot bind: <1,1> type:users (ref<1,2>, => <2,0>)
   4: list bind: <1,3> type:[]user
   5: $all bind: <1,4> type:bool
   6: {} bind: <1,4> type:[]block (=> <3,0>)
-> block 2
   entrypoints: [<2,7>]
   1: user id = context
   2: sshkeys bind: <2,1> type:[]privatekey
   3: length bind: <2,2> type:int
   4: > bind: <2,3> type:bool (2)
   5: enabled bind: <2,1> type:bool
   6: == bind: <2,5> type:bool (true)
   7: && bind: <2,4> type:bool (ref<2,6>)
-> block 3
   entrypoints: [<3,2> <3,3> <3,4> <3,9>]
   1: user id = context
   2: name bind: <3,1> type:string
   3: uid bind: <3,1> type:int
   4: gid bind: <3,1> type:int
   5: sshkeys bind: <3,1> type:[]privatekey
   6: length bind: <3,5> type:int
   7: > bind: <3,6> type:bool (2)
   8: enabled bind: <3,7> type:bool
   9: == bind: <3,8> type:bool (true)
```

block 3, chunk 8 seems incorrect to me: <3,7> is of type bool. We are calling `enabled` on it

Reply:
Thanks for catching this bug. I have to look into this. For some queries, we only need the last chunk as entrypoint. This prevents, e.g, for `.all(sshkeys.length == 0)` the complete data for `sshkeys` to show up. Otherwise we would see `.length` and
```
sshkeys: [
        0: user privatekey id = privatekey:/home/christian/.ssh/xyz
...
      ]
```
But in the example you found, the data for the 2nd part after `&&` is missing completely. This should show up.

Reply:
I opted for the solution you suggested. I couldn't find a way to make this work with all sorts of queries and, at the same time, keep the output minimal.

The query you mentioned now also works, and also some more. But we display more data. I don't know whether all this data is needed. I'll keep this in the back of my head. Maybe I'll find a solution in a follow-up PR.